[lldb/Interpreter] Fix ambiguous partial command resolution

medismailben opened the pull request:
This patch is a follow-up to #97263 that fix ambigous abbreviated
command resolution.

When multiple commands are resolved, instead of failing to pick a command to
run, this patch changes to resolution logic to check if there is a single
alias match and if so, it will run the alias instead of the other matches.

This has as a side-effect that we don't need to make aliases for every
substring of aliases to support abbrivated alias resolution.

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>

Michael137 commented on lldb/source/Interpreter/CommandInterpreter.cpp:
```diff
@@ -1302,6 +1298,36 @@ CommandObject *CommandInterpreter::GetUserCommandObject(
 
+CommandObject *CommandInterpreter::GetAliasCommandObject(
+    llvm::StringRef cmd, StringList *matches, StringList *descriptions) const {
+  std::string cmd_str(cmd);
+  auto find_exact = [&](const CommandObject::CommandMap &map) {
+    auto found_elem = map.find(std::string(cmd));
```
```suggestion
    auto find_exact = [&](const CommandObject::CommandMap &map) {
    auto found_elem = map.find(cmd.str());
```
(in C++20 we could just pass the StringRef to `find`, but not sure that's the default version yet)

medismailben replied:
Not yet unfortunately.

Michael137 commented on lldb/source/Interpreter/CommandInterpreter.cpp:
```diff
@@ -1302,6 +1298,36 @@ CommandObject *CommandInterpreter::GetUserCommandObject(
+  // We didn't have an exact command, so now look for partial matches.
+  StringList tmp_list;
+  StringList *matches_ptr = matches ? matches : &tmp_list;
+  AddNamesMatchingPartialString(GetAliases(), cmd_str, *matches_ptr);
```
```suggestion
  AddNamesMatchingPartialString(GetAliases(), cmd, *matches_ptr);
```

Michael137 commented on lldb/source/Interpreter/CommandInterpreter.cpp:
```diff
@@ -1302,6 +1298,36 @@ CommandObject *CommandInterpreter::GetUserCommandObject(
+    if (found_elem == map.end())
+      return (CommandObject *)nullptr;
+    CommandObject *exact_cmd = found_elem->second.get();
+    if (exact_cmd) {
+      if (matches)
+        matches->AppendString(exact_cmd->GetCommandName());
+      if (descriptions)
+        descriptions->AppendString(exact_cmd->GetHelp());
+      return exact_cmd;
+    }
+    return (CommandObject *)nullptr;
```
```suggestion
    if (!exact_cmd)
      return nullptr;
      
    if (matches)
      matches->AppendString(exact_cmd->GetCommandName());
  
    if (descriptions)
      descriptions->AppendString(exact_cmd->GetHelp());
    
    return exact_cmd;
```

Michael137 commented on lldb/source/Interpreter/CommandInterpreter.cpp:
```diff
@@ -1302,6 +1298,36 @@ CommandObject *CommandInterpreter::GetUserCommandObject(
+CommandObject *CommandInterpreter::GetAliasCommandObject(
+    llvm::StringRef cmd, StringList *matches, StringList *descriptions) const {
+  std::string cmd_str(cmd);
+  auto find_exact = [&](const CommandObject::CommandMap &map) {
```
Nit:
```suggestion
  auto find_exact = [&](const CommandObject::CommandMap &map) -> CommandObject * {
```
Then you don't need the casts below.

jimingham commented:
IIUC, you are adding a new rule that if an incoming command name has partial matches to one user command and one alias command, the user command is preferred over the alias command.  Is that right?

You certainly should document that rule somewhere - maybe in the Tutorial section of the docs?  Maybe in the help for `command alias`.  And there should certainly be an explicit test to show this ordering - you can just use the script/scripting since that's what you added this for.

But I think I'm missing something, because I don't see how this solves the problem you set out to solve.  You wanted to add the alias `script` to point to `scripting run`, and make `sc`, `scr`, etc also mean your alias.  However, if I'm following your logic correctly, if I type `scr` then that matches the Alias command `script` -> `scripting run` and the user command `scripting`.  By the logic in your path, it will choose the user command `scripting` not the alias, which I think is the opposite of what you wanted.

I must be missing something.

medismailben commented:
> IIUC, you are adding a new rule that if an incoming command name has partial matches to one user command and one alias command, the user command is preferred over the alias command. Is that right?

Yes.

> You certainly should document that rule somewhere - maybe in the Tutorial section of the docs? Maybe in the help for `command alias`. And there should certainly be an explicit test to show this ordering - you can just use the script/scripting since that's what you added this for.

Indeed, I should document that.

> But I think I'm missing something, because I don't see how this solves the problem you set out to solve. You wanted to add the alias `script` to point to `scripting run`, and make `sc`, `scr`, etc also mean your alias. However, if I'm following your logic correctly, if I type `scr` then that matches the Alias command `script` -> `scripting run` and the user command `scripting`. By the logic in your path, it will choose the user command `scripting` not the alias, which I think is the opposite of what you wanted.
>
> I must be missing something.

Right, but I think commands added explicitly by the user, that overwrite other commands (or multiword commands in the `scripting` example), should have priority over the lldb defined commands and aliases.

medismailben commented:
> This is for this sort of lookup:
> 
> ```
> (lldb) reg read pc
> ```
> 
> Right? As opposed to tab completion.

Right, tab completion already can suggest multiple candidates, so that worked already.

> This needs a test case or at least a clear example as a comment in the code if a test is not possible. It seems fine from how you've described it but these things come across better with a clear example.

Sorry if this wasn't clear, but I didn't write a test case because there was already a test case in https://github.com/llvm/llvm-project/pull/97263, but I guess I could write a test to test the ordering precisely (custom command, then alias, then native command)

jimingham commented:
> > IIUC, you are adding a new rule that if an incoming command name has partial matches to one user command and one alias command, the user command is preferred over the alias command. Is that right?
> 
> Yes.
> 
> > You certainly should document that rule somewhere - maybe in the Tutorial section of the docs? Maybe in the help for `command alias`. And there should certainly be an explicit test to show this ordering - you can just use the script/scripting since that's what you added this for.
> 
> Indeed, I should document that.
> 
> > But I think I'm missing something, because I don't see how this solves the problem you set out to solve. You wanted to add the alias `script` to point to `scripting run`, and make `sc`, `scr`, etc also mean your alias. However, if I'm following your logic correctly, if I type `scr` then that matches the Alias command `script` -> `scripting run` and the user command `scripting`. By the logic in your path, it will choose the user command `scripting` not the alias, which I think is the opposite of what you wanted.
> > I must be missing something.
> 
> Right, but I think commands added explicitly by the user, that overwrite other commands (or multiword commands in the `scripting` example), should have priority over the lldb defined commands and aliases.

I'm not sure I agree with this.  If I'm used to typing `br` for break, and then someone adds `brogue` (speaks the result of an expression in a Scottish accent?) I would be very surprised if `br` started meaning `brogue`.  It would be much better to tell me there was now a conflict, and I can figure out what I want to do about it.

However, aliases are going to be the way I resolve this issue for myself (by adding an alias for `br` that goes to `break`, for instance.  So it seems okay, if well explained, to have aliases silently win over built-in and user commands, I don't think it's a good idea for user commands to do so over built-in commands.

medismailben commented:
> However, aliases are going to be the way I resolve this issue for myself (by adding an alias for `br` that goes to `break`, for instance.) So while it seems okay, if well explained, to have aliases silently win over built-in and user commands, I don't think it's a good idea for user commands to do so over built-in commands.

Ok, I don't have a strong opinion about user commands so we ditch them from this new logic.

jimingham commented:
Im my mind, user commands and built-in commands are almost identical, the only difference being currently we don't allow you to delete built-in commands.  So in cases like this it makes more sense to treat user & built-in commands the same way as far as lookup and parsing.

We should also document this lookup rule (in `help command alias` and in the Tutorial where it describes the basic command line parsing).

medismailben commented:
@Michael137 @jimingham I've addressed your comment. The test is part of the https://github.com/llvm/llvm-project/pull/97263.

jimingham commented:
It does seem worthwhile to have a test explicitly testing this ordering that exists among the tests for the command interpreter, rather than having it be a side-effect test in an unrelated test case.

Other than that, this is fine.

medismailben committed:
[lldb/Interpreter] Fix ambiguous partial command resolution

This patch is a follow-up to #97263 that fix ambigous abbreviated
command resolution.

When multiple commands are resolved, instead of failing to pick a command to
run, this patch changes to resolution logic to check if there is a single
alias match and if so, it will run the alias instead of the other matches.

This has as a side-effect that we don't need to make aliases for every
substring of aliases to support abbrivated alias resolution.

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>

jimingham reviewed (approved):
LGTM
